File: Projeto/mpf.py
import math

def fixed_point_iteration(g, x0, tol=1e-6, max_iter=100):
    """
    Método do Ponto Fixo para encontrar a raiz de f(x) = 0,
    usando uma função g(x) tal que x = g(x).

    :param g: Função g(x) usada para a iteração.
    :param x0: Valor inicial (chute).
    :param tol: Tolerância para o erro.
    :param max_iter: Número máximo de iterações.
    :return: A raiz aproximada, número de iterações e lista de aproximações.
    """
    iter_count = 0
    x = x0
    values = [x0]

    while iter_count < max_iter:
        x_next = g(x)
        values.append(x_next)

        if abs(x_next - x) < tol:
            return x_next, iter_count + 1, values

        x = x_next
        iter_count += 1

    raise Exception("O método não convergiu dentro do número máximo de iterações.")

if __name__ == "__main__":
    g = lambda x: math.sqrt(6-x)
    x0 = 2.25

    try:
        root, iters, history = fixed_point_iteration(g, x0)
        print(f"Raiz aproximada: {root}")
        print(f"Número de iterações: {iters}")
        print("Histórico de aproximações:")
        for i, val in enumerate(history):
            print(f"  Iteração {i}: {val}")
    except Exception as e:
        print(e)

File: Projeto/haskell/mpf.hs
import Prelude
import Data.Maybe (fromMaybe)
import System.IO
import GHC.Float (double2Float)
import Prelude
import Text.Printf (printf)

-- Método de iteração de ponto fixo que retorna a lista de iterações e status de convergência
fixedPoint :: (Double -> Double) -> Double -> Double -> Int -> ([Double], Bool)
fixedPoint g x0 tol maxIter = (go x0 0, converged x0 0)
  where
    go x n
      | n >= maxIter         = [x]
      | abs (xNext - x) < tol = [x, xNext]
      | otherwise            = x : go xNext (n + 1)
      where
        xNext = g x

    converged x n
      | n >= maxIter          = False
      | abs (xNext - x) < tol = True
      | otherwise             = converged xNext (n + 1)
      where
        xNext = g x

-- Exemplo: resolver x = (x^3 + 3) / 9
main :: IO ()
main = do
    let g = \x -> (x ** 3 + 3) / 9
        x0 = 0.5
        tol = 1e-3
        maxIter = 100
        (iters, didConverge) = fixedPoint g x0 tol maxIter

    putStrLn "Iterations:"
    mapM_ (\(i, x) -> printf "Iter %d: %.8f\n" i x) (zip [0 :: Int ..] iters)

    if didConverge
      then printf "Root found: %.8f\n" (last iters)
      else putStrLn "Did not converge."